services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala

```diff
@@ -13,38 +13,73 @@ import scala.concurrent.{ ExecutionContext, Future }
 class GitLabProjectVersioning(httpClient: HttpClient, config: GitLabConfig)
     extends ProjectVersioning[VersioningException] {
 
-  override def updateFile(project: Project, projectFile: ProjectFile, version: Option[Version])(
+  override def updateFile(project: Project, projectFile: ProjectFile, version: Option[PipelineVersion])(
     implicit ec: ExecutionContext
   ): AsyncResult[String] = {
     val path = URLEncoder.encode(projectFile.path.toString, "UTF-8")
-    val fileUrl = s"${config.url}projects/${project.repository}/repository/files/$path"
-    val versionValue = version.map(_.name).getOrElse(config.defaultFileVersion)
+    val repositoryId: Repository = project.repository match {
+      case Some(repository) => repository
+      case None             => throw VersioningException(s"No repository for project: $project")
+    }
+    val fileUrl = s"${config.url}projects/${repositoryId.value}/repository/files/$path"
+
+    getProjectVersions(project).flatMap(versions => {
+      val latestVersion: Either[VersioningException, PipelineVersion] = (version, versions) match {
+        case (Some(_), Left(error)) => Left(error)
+        case (Some(userVersion), Seq(headVersion: GitLabVersion, _)) if headVersion.name >= userVersion =>
```

Comment:
If I'm not mistaken, you are comparing here strings instead of versions. I.e. you are not using `compare` method that you implemented.

Reply:
I'm kinda surprised that it compiled, but it looks like the compiler did try to stop you. That's why you forced him to think that `headVersion` is a `GitLabVersion` instead of `Any`.
`versions` is an `Either`. If you want to pattern match on it, you must write `Right(Seq(...))`.
Same for line 36 below.

Reply:
Ok, thx

Reply:
About not using compare method: `headVersion` is `GitLabVersion` type, that has param `name`  of `PipelineVersion` type